pkg/verifier/safe_delete.go

```diff
@@ -31,23 +36,45 @@ func SafeDelete(ctx context.Context, logger log.Logger, bkt objstore.Bucket, bac
 		return errors.Errorf("%s dir seems to exists in backup bucket. Remove this block manually if you are sure it is safe to do", id)
 	}
 
-	tempdir, err := ioutil.TempDir("", "safe-delete")
-	if err != nil {
-		return err
+	if tempdir == "" {
+		// tempdir unspecified, create one
+		tempdir, err = ioutil.TempDir("", fmt.Sprintf("safe-delete-%s", id.String()))
+		if err != nil {
+			return err
+		}
+
+		// We manage this tempdir so we clean it up on exit.
+		defer func() {
+			if err := os.RemoveAll(tempdir); err != nil {
+				level.Warn(logger).Log("msg", "failed to delete dir", "dir", tempdir, "err", err)
+			}
+		}()
 	}
 	dir := filepath.Join(tempdir, id.String())
-	err = os.Mkdir(dir, 0755)
-	if err != nil {
+
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		// TSDB block not already present, create it.
+		err = os.Mkdir(dir, 0755)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		// Error calling Stat() and something is really wrong.
 		return err
 	}
-	defer func() {
-		if err := os.RemoveAll(tempdir); err != nil {
-			level.Warn(logger).Log("msg", "failed to delete dir", "dir", tempdir, "err", err)
-		}
-	}()
 
-	if err := block.Download(ctx, logger, bkt, id, dir); err != nil {
-		return errors.Wrap(err, "download from source")
+	if useExisting {
+		if _, err := os.Stat(filepath.Join(dir, "meta.json")); err != nil {
```

Comment:
Is this check really sufficient? Shouldn't we check that all the downloaded content is equal to what is in object storage, ideally by their hash?

Reply:
There don't appear to be tools for this in the existing code base.  So this seems like a bit of scope creep.  Glad to add additional checks here, but these features sound like another GitHub Issue/PR.

In fact, why isn't there a checksum in meta.json?  It doesn't have to be cryptographically secure, but it would serve to verify that the block was intact.  As we move from "metric data is ephemeral" to "metric data is important long term" that would be quite handy.